### PR TITLE
geometry caching

### LINKED ISSUE #682: Add `fix_face` and include when faces are created
## Description of issue / requirement to address

We've come across an issue in #589 in which a `BluemiraFace` is created from an allegedly valid `BluemiraWire` but the resulting face is not valid. It was found that calling `.fix(float, float, float)` on this fixed the face.

## Proposed solution

Add a `fix_face` function in `_freecadapi.py` and use it (under the hood) when `BluemiraFace`s are created. Warn if the resulting face required fixing.

I am not proposing to make this a user-facing function for the time being.

## Alternative solutions

Haven't come across any good alternatives. Given that this can happen with a wire that is supposedly valid means we will need some form of detection and protection against these cases sooner or later.

### LINKED ISSUE #1090: Faling build of hollow torus solid
## Describe the bug

I am trying to build a hollowed torus of radii `a_in` and `a_out` and major radius `R_0` by making a two concentring circular wires in the xz plane, get the corresonding hollowed face and revolve it by 360.0 degree.

I get some unexpected error

## Steps to reproduce

Simply run the following function

``` python
from bluemira.base.components import Component, PhysicalComponent
from bluemira.geometry.face import BluemiraFace
from bluemira.geometry.tools import make_circle, revolve_shape, wire_closure

def make_torus(a_in: float, a_out: float, R_0: float) -> PhysicalComponent:
    """
    create a Physical component containing a hollowed torus geometry
    of major radius R_0 and inner and outer minor radius a_1 and a_2
    respectively
    """
    # Make (open) initial wire
    inner_circle = make_circle(
        radius=a_in,
        center=(R_0, 0.0, 0.0),
        axis=(0.0, 1.0, 0.0),
        label="in_torus"
    )
    outer_circle = make_circle(
        radius=a_out,
        center=(R_0, 0.0, 0.0),
        axis=(0.0, 1.0, 0.0),
        label="out_torus"
    )

    # # close initial circles
    # inner_circle = wire_closure(inner_circle, label="in_torus")
    # outer_circle = wire_closure(outer_circle, label="out_torus")
    # print("DONKEY")

    # Make the face
    face = BluemiraFace([outer_circle, inner_circle], label="face")
    torus = revolve_shape(shape=face, degree=360.0)
    
    return PhysicalComponent("Little torus", torus)
```

## Expected behaviour

A nice torus component

## Screenshots / Tracebacks

``` bash
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/skahn/codes/step/step-blueprint-reactor-design/tests/test_builder_tools.py", line 38, in make_torus
    return PhysicalComponent("Little torus", torus)
  File "/home/skahn/codes/bluemira/bluemira/geometry/tools.py", line 464, in wire_closure
    closure = BluemiraWire(cadapi.wire_closure(wire), label=label)
  File "/home/skahn/codes/bluemira/bluemira/geometry/wire.py", line 65, in __init__
    super().__init__(boundary, label, boundary_classes)
  File "/home/skahn/codes/bluemira/bluemira/geometry/base.py", line 97, in __init__
    self.boundary = boundary
  File "/home/skahn/codes/bluemira/bluemira/geometry/base.py", line 146, in boundary
    self._boundary = self._check_boundary(objs)
  File "/home/skahn/codes/bluemira/bluemira/geometry/base.py", line 133, in _check_boundary
    raise TypeError(
TypeError: Only [<class 'bluemira.geometry.wire.BluemiraWire'>, <class 'Part.Wire'>] objects can be used for <class 'bluemira.geometry.wire.BluemiraWire'>
```

Trying to close the wires, I get using `wire_close`, I get 

``` bash
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/skahn/codes/step/step-blueprint-reactor-design/tests/test_builder_tools.py", line 30, in make_torus
    inner_circle = wire_closure(inner_circle, label="in_torus")
  File "/home/skahn/codes/bluemira/bluemira/geometry/tools.py", line 464, in wire_closure
    closure = BluemiraWire(cadapi.wire_closure(wire), label=label)
  File "/home/skahn/codes/bluemira/bluemira/geometry/wire.py", line 65, in __init__
    super().__init__(boundary, label, boundary_classes)
  File "/home/skahn/codes/bluemira/bluemira/geometry/base.py", line 97, in __init__
    self.boundary = boundary
  File "/home/skahn/codes/bluemira/bluemira/geometry/base.py", line 146, in boundary
    self._boundary = self._check_boundary(objs)
  File "/home/skahn/codes/bluemira/bluemira/geometry/base.py", line 133, in _check_boundary
    raise TypeError(
TypeError: Only [<class 'bluemira.geometry.wire.BluemiraWire'>, <class 'Part.Wire'>] objects can be used for <class 'bluemira.geometry.wire.BluemiraWire'>
```

### PR DESCRIPTION
**Pull request by ivanmaione**
## Linked Issues

related to Issue #1196
Closes #682 
Closes #1090

## Description

This PR implements the shape caching in BluemiraGeo objects. Geometry functions are modified in such a way that shape and boundary are always consistent.

## Interface Changes

BluemiraGeo.shape and BluemiraGeo.boundary are readonly now.

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

## Check tests
tests/geometry
- [x] test_coordinates.py
- [x] test_deprecated_base.py
- [x] test_deprecated_loop.py
- [x] test_deprecated_offset.py
- [x] test_deprecated_tools.py
- [x] test_face.py
- [x] test_geometry.py
- [x] test_inscribed_rect.py
- [x] test_offset.py -> necessary to change test_data/offset_wire2022-04-08_10-19-27.json to be compliant with the new serialization structure.
- [x] test_optimisation.py
- [x] test_parameterisations.py
- [x] test_orientation.py -> since the FreeCAD shape is directly cached, the only orientation test that could have meaning is that in which two wires are joined -> maybe it would be worth discussing this point.
- [x] test_placement.py
- [x] test_plane.py
- [x] test_sweep.py
- [x] test_tools.py
- [x] test_wire.py

### CODE REVIEW

File: bluemira/geometry/base.py [hunk 1]
```diff
@@ -94,9 +94,9 @@ def __init__(
     ):
         super().__init__()
         self._boundary_classes = boundary_classes
-        self.boundary = boundary
-        self.label = label
         self.__orientation = _Orientation("Forward")
```
Comment (CoronelBuendia) [inline bluemira/geometry/base.py:97]: `_orientation` and `__orientation` both exist. I don't think this is intended

File: bluemira/geometry/solid.py [hunk 2]
```diff
@@ -61,27 +64,82 @@ def _create_solid(self, check_reverse=True):
...
                 bm_shells.append(BluemiraShell._create(shell))
 
-            bmsolid = cls(bm_shells, label=label)
-            bmsolid._orientation = orientation
+            # create an empty BluemiraSolid
+            bmsolid = cls(None, label=label)
+            # assign shape, boundary, and orientation
+            bmsolid._set_shape(obj)
+            bmsolid._boundary = bm_shells
+            bmsolid._orientation = obj.Orientation
```
Comment (CoronelBuendia) [inline bluemira/geometry/solid.py:91]: E.g. here

File: bluemira/geometry/tools.py [hunk 3]
```diff
@@ -813,9 +813,7 @@ def split_wire(wire: BluemiraWire, vertex: Iterable, tolerance: float = EPS):
     GeometryError:
         If the vertex is further away to the wire than the specified tolerance
     """
-    wire_1, wire_2 = cadapi.split_wire(
-        wire.get_single_wire()._shape, vertex, tolerance=tolerance
-    )
+    wire_1, wire_2 = cadapi.split_wire(wire.shape, vertex, tolerance=tolerance)
```
Comment (CoronelBuendia) [inline bluemira/geometry/tools.py:816]: Here you change to `.shape` but in many of the other functions in tools.py, they still use `._shape`. Is there a reason it is different here?

File: examples/geometry/geometry_test.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+# bluemira is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+
+"""
+Application test for the caching geoemtry (ivanmaione/develop_simple_caching_2)
```
Comment (CoronelBuendia) [inline examples/geometry/geometry_test.py:23]: As discussed, to be merged with the other geometry example and removed (including ipynb version).

File: bluemira/builders/cryostat.py [hunk 5]
```diff
@@ -148,15 +148,15 @@ def build_xz(self, x_out: float, z_top: float) -> PhysicalComponent:
...
+        z_inner = np.array([z_top, z_top, z_mid, z_mid, z_bot, z_bot])
 
         x_outer = np.array([x_in, x_gs_kink, x_gs_kink, x_out, x_out, x_in])
         x_outer[1:-1] += tk
 
         z_outer = np.array([z_bot, z_bot, z_mid, z_mid, z_top, z_top])
-        z_outer[:4] -= tk
-        z_outer[4:] += tk
+        z_outer[:2] -= tk
+        z_outer[2:] += tk
```
Comment (CoronelBuendia) [inline bluemira/builders/cryostat.py:159]: Strange... i trust it works now, but when we come to redoing the EUDEMO stuff I will look again

**Review by CoronelBuendia — CHANGES_REQUESTED** [review]
Looking really good! I fixed the last remaining test failure and made a couple of minor aesthetic changes. A couple of comments / queries beyond what we had discussed.

I'd really like to get rid of the orientation stuff in general, but without the EU-DEMO build stuff running in full with the caching, I'm too scared to take it out. I will create a separate issue to remove the orientation tracking, now that we no longer handle reconstruction.

File: examples/geometry/geometry_tutorial.py [hunk 6]
```diff
@@ -68,54 +72,132 @@
...
 
 # %%[markdown]
+# ## Geometry creation (1-D)
+
+# Let's get familiar with some ways of making 1-D geometries.
+# Bluemira implements functions for the creation of:
+# # * polygons
+# # * splines
+# # * arcs
+# # * a bit of everything (check geometry.tools module for an extensive list)
```
Comment (je-cook) [inline examples/geometry/geometry_tutorial.py:85]: Some extra hashes
```suggestion
# * polygons
# * splines
# * arcs
# * a bit of everything (check geometry.tools module for an extensive list)
```

Comment (je-cook) [inline examples/geometry/geometry_tutorial.py:85]: in future we should probably link to things like this. Its easier for a user to just click on the link. Not blocking for now and probably part of a larger discussion when we do the docs review.

File: bluemira/geometry/base.py (hunk 1 above)
Comment (je-cook) [reply bluemira/geometry/base.py:97]: I agree this should always go through the setter of `_orientation`

**Review by je-cook — APPROVED** [review]
I'm happy just a few minors

File: bluemira/geometry/base.py [hunk 7]
```diff
@@ -122,14 +122,23 @@ def _converter(func):
...
         check = False
         for c in self._boundary_classes:
-            check = check or (all(isinstance(o, c) for o in objs))
+            # # in case of obj = [], this check returns True instead of False
+            # check = check or (all(isinstance(o, c) for o in objs))
+            for o in objs:
+                check = check or isinstance(o, c)
             if check:
-                return objs
+                return tuple(objs)
```
Comment (ivanmaione) [inline bluemira/geometry/base.py:141]: @CoronelBuendia If it does not disturb the fact that self._boundary is a list and self.boundary returns a tuple, I would propose converting the list into a tuple in the boundary getter (in this way it is still allowed to do self._boundary.append(things) as in some file).

**Review by CoronelBuendia — APPROVED** [review]

**Pull request merged**